tskv/src/compaction/picker.rs

```diff
@@ -26,38 +43,211 @@ impl Picker for LevelCompactionPicker {
         tsf_id: TseriesFamilyId,
         version: Arc<Version>,
     ) -> Option<CompactReq> {
+        //! 1. Get a TseriesFamily (by TseriesFamily itself), and it's **version**(`Arc<Version>`),
+        //!    **ts_family_id**, and ts_family_opt(`Arc<TsFamilyOptions>)
+        //! 2. Get time_range duration: `${config.time_window}` (by TseriesFamOpt).
+        //! 3. Get rounded time_range for specified timestamp ( by function
+        //!    rounded_time_range(`${args.timestamp}`, `${config.time_window}`) )
+        //! 4. Get levels(`Vec<LevelInfo>`) for the time_range (by LevelInfo::time_range)
+        //! 5. Get files(`Vec<Arc<ColumnFile>>`) from the level (by ColumnFile::time_range),
+        //!    sorted by min_ts(ascending) and size(ascending).
+        //! 6. Get the max level in the selected levels, and get the **out_level** (max_level + 1)
+        //! 7. Get parameter **max_compact_size** for the out_level.
+        //! 8. Add each `Arc<ColumnFile>` to **files**, until the total size is bigger than max_compact_size.
+        //! 9. Build CompactReq using **ts_family_id**, **ts_family_opt**,
+        //!    **version**, **files**, **out_level**
+
         let level_infos = version.levels_info();
 
-        let mut ctx = LevelCompatContext::default();
-        ctx.cal_score(level_infos, self.ts_family_opt.as_ref());
-        if let Some((start_level, out_lvl)) = ctx.pick_level() {
-            let input = ctx.pick_files(
-                level_infos,
-                self.ts_family_opt.as_ref(),
-                start_level,
-                out_lvl,
-            );
-            if let Some((level, mut files)) = input {
-                for file in files.iter_mut() {
-                    file.mark_compaction();
-                }
-                let request = CompactReq {
-                    ts_family_id: tsf_id,
-                    ts_family_opt: self.ts_family_opt.clone(),
-                    files,
-                    version,
-                    out_level: out_lvl,
-                };
-                return Some(request);
+        // Pick a level to compact with level 0
+        let level_start: &LevelInfo;
+        let out_level;
+
+        if let Some((start_lvl, out_lvl)) = self.pick_level(level_infos) {
+            level_start = &level_infos[start_lvl as usize];
+            out_level = out_lvl;
+        } else {
+            return None;
+        }
+        let max_compact_size = self.ts_family_opt.level_file_size(out_level);
+
+        // Pick selected level files.
+        let mut picking_files: Vec<Arc<ColumnFile>> = Vec::new();
+        let mut picking_files_size = 0_u64;
+        let mut picking_time_range = TimeRange::from((Timestamp::MAX, Timestamp::MIN));
+
+        if level_start.files.len() >= 1 {
```

Comment:
is empty

Reply:
condition `length >= 1` already contained `not empty`.